models/cost_per_serverless_task_run.sql

```diff
@@ -0,0 +1,151 @@
+    unique_key=['query_id', 'start_time'],
+) }}
+
+with
+stop_threshold as (
+    select max(end_time) as latest_ts
+    from {{ ref('stg_serverless_task_history') }}
+),
+
+// task_history has all runs, but we only want serverless runs
+serverless_runs as (
+    select
+        task_history.* exclude (query_id), // just to remove ambiguous column names
+        query_history.*,
+        task_history.database_name as task_database_name,
+        task_history.schema_name as task_schema_name,
+        name as task_name,
+        concat(task_database_name, '.', task_schema_name, '.', task_name) as task_fqn_name,
+        start_time as execution_start_time,
+    from {{ ref('stg_task_history') }} as task_history
+
+    inner join {{ ref('stg_query_history') }} as query_history
+        on task_history.query_id = query_history.query_id and query_history.start_time = task_history.query_start_time
+
+    where true
+        and task_schema_id is not null
+        and task_database_id is not null
+        and startswith(warehouse_name, 'COMPUTE_SERVICE_WH_USER_TASKS') -- TODO: Verify on a larger sample
+        and end_time <= (select latest_ts from stop_threshold)
+        {% if is_incremental() %}
+        -- account for late arriving queries
+        and end_time > (select coalesce(dateadd(day, -3, max(end_time)), '1970-01-01') from {{ this }})
+        {% endif %}
+
+    order by scheduled_time desc
+),
+
+hours_list as (
+    select
+        dateadd(
+            'hour',
+            '-' || row_number() over (order by seq4() asc),
+            dateadd('day', '+1', current_date::timestamp_tz)
+        ) as hour_start,
+        dateadd('hour', '+1', hour_start) as hour_end
+
+    {% if is_incremental() %}
+    from table(generator(rowcount => (24 * 7)))
+    {% else %}
+    from table(generator(rowcount => (24 * 730)))
+    {% endif %}
+),
+
+-- 1 row per hour a run ran
+run_hours as (
+    select
+        hours_list.hour_start,
+        hours_list.hour_end,
+        runs.*
+    from hours_list
+    inner join serverless_runs as runs
+        on hours_list.hour_start >= date_trunc('hour', runs.execution_start_time)
+            and hours_list.hour_start < runs.end_time
+),
+
+run_seconds_per_hour as (
+    select
+        *,
+        datediff('millisecond', greatest(execution_start_time, hour_start), least(end_time, hour_end)) as num_milliseconds_query_ran,
+        sum(num_milliseconds_query_ran) over (partition by task_fqn_name, hour_start) as total_query_milliseconds_in_hour,
+        div0(num_milliseconds_query_ran, total_query_milliseconds_in_hour) as fraction_of_total_query_time_in_hour,
+        hour_start as hour
+    from run_hours
+),
+
+credits_billed_hourly as (
+    select
+        start_time as hour,
+        database_name,
+        database_id,
+        schema_name,
+        schema_id,
+        task_name,
+        credits_used as credits_used_serverless_task
+    from {{ ref('stg_serverless_task_history') }}
+    group by all
```

Comment:
should we do:

```sql
    select
        start_time as hour,
        database_id,
        schema_id,
        task_name,
        sum(credits_used) as credits_used_serverless_task
    from {{ ref('stg_serverless_task_history') }}
    group by all
```

to enforce the desired grain? we later join on `database_id, schema_name, task_id` so not sure why we need the other columns kicking around?

Reply:
Good idea. I added the database and schema names when I was troubleshooting it, but indeed we only use the IDs. I just removed the other fields, thanks!